bin/jsii-release-golang

```diff
@@ -56,68 +72,121 @@ if [ -z "${GITHUB_TOKEN:-}" ]; then
+# Otherwise, reads the `version` file in the module directory.
+# Fails if none exist.
+function extract_version {
+  module=$1
+  if [ ! -z "${VERSION:-}" ]; then
+    module_version=${VERSION}
+  elif [ -f ${module}/version ]; then
+    module_version=$(cat ${module}/version)
+  else
+    error "Unable to determine version of module ${module}. Either include a 'version' file, or specify a global version using the VERSION environment variable."
+  fi
 
-if [ ${go_mod} = "" ]; then
-  error "No go modules found in ${dir}"
-fi
+  echo ${module_version}
+}
+
+# Extracts the repository name from the collection of modules
+# by parsing the go.mod file of one of the modules, assuming all modules are pushed
+# to the same repository.
+function extract_repo {
+  modules=$1
+  go_mod=${modules[0]}/go.mod
+  full_module_name=$(grep module ${go_mod} | cut -d ' ' -f 2)
+  owner=$(echo ${full_module_name} | cut -d '/' -f 2)
+  repo=$(echo ${full_module_name} | cut -d '/' -f 3)
+  echo ${owner}/${repo}
+}
+
+# Creates the necessary tag for a given module.
+# If the module is located at the root directory, the tag will in the form of `v${module_version}`.
+# Otherwise, the module name is prefixed to it, e.g `my-module/v${module_version}`.
+function create_tag {
+  module=$1
+
+  module_name=$(basename ${module})
+  module_version=$(extract_version ${module})
+
+  if [ ${module} = ${dir} ]; then
+    # root module
+    tag_name=v${module_version}
+  else
+    # submodule
+    tag_name=${module_name}/v${module_version}
+  fi
 
-unique_versions=($(echo "${versions[@]}" | tr ' ' '\n' | sort -u | tr '\n' ' '))
-count_unique_version=$(echo "${#unique_versions[@]}")
+  git tag -a ${tag_name} -m ${tag_name}
+  if $dry_run; then
+    echo "Will push tag: ${tag_name}"
+  else
+    git push origin ${tag_name}
+  fi
+}
+
+# Delete all modules in the target repository in preparation of
+# syncing it with the source code located in `DIR`.
+function purge_modules {
+  repo_dir=$1
+
+  # first remove all submodules
+  for subdir in $(find ${repo_dir} -mindepth 1 -maxdepth 1 -type d); do
+    if [ -f ${subdir}/go.mod ]; then
+      rm -r ${subdir}
+    fi
+  done
+
+  # if the root is a module too, remove the entire repo.
+  if [ -f ${dir}/go.mod ]; then
+    rm -r ${repo_dir}/*
+  fi
 
-if [ ${count_unique_version} -gt 1 ]; then
-  error "multiple versions are not supported. please use a single version across all your modules."
-fi
+}
 
-if [ -z "${VERSION:-}" ]; then
+modules=()
 
-  if [ ${count_unique_version} -eq 0 ]; then
-    error "No version files found. Please supply a version using the VERSION env variable."
-  fi
+# collect mod files from root and subdirectories
+for go_mod in $(find ${dir} -name go.mod -mindepth 1 -maxdepth 2); do
+  modules+=($(dirname ${go_mod}))
+done
 
-  # must be a single element at this point
-  VERSION=${unique_versions[0]}
+# at least one go module must exist
+if [ ${#modules[@]} -eq 0 ]; then
+  error "No go modules detected in ${dir}"
```

Comment:
What's the behavior of other publishers? I vaguely remember this is a no-op (and not an error)

Reply:
Yeah, switched to no-op